[string] Fix string implementation for big endian platforms

mundaym opened the pull request:
This commit supersedes 2866b4a which was overwritten by 4ab45df.

When code points are stored in multi-byte types (e.g. UInt64) we
need to convert the order of the bytes in the value to little
endian before we can access the value as raw bytes.

milseman commented on stdlib/public/core/SmallString.swift:
```diff
@@ -210,7 +204,10 @@ extension _SmallString {
+    // Code points need to be in little endian byte order and zero terminated.
+    let l = self.leadingRawBits
+    let t = self.trailingRawBits & _StringObject.Nibbles.largeAddressMask
+    var raw = (l.littleEndian, t.littleEndian)
```
Why do we convert the bytes to little endian on read? If the plan is to store the contents in big-endian on big-endian systems, then we don't want to convert on read, right?

milseman replied:
And if not, then would it make more sense to just have a `littleEndianBytes` computed property?

mundaym replied:
The trouble is that the discriminator is stored in the most significant byte of the `_object` pointer. On little endian systems this means the discriminator is stored in the 'last' byte of the string objects memory:
```
[ x x x x x x x x x x x x x x x d ]
```
However on big endian systems this means that the discriminator is stored in the middle byte of the string object:
```
[ x x x x x x x x d x x x x x x x ]
```
Since the discriminator is in the middle we can only access the first 8 code units with a `withUnsafeBytes` call if we use big endian byte order. However if we byte swap to little endian byte order before any `withUnsafeBytes` calls then we can access all 15 code units and so big endian machines can do all the same operations a little endian machine can do. It also means we can add code units to the small string in the same order on both big and little endian platforms which is much cleaner and easier to manage.

We could add a `littleEndianBytes` computed property but I don't think the code saving is worth the additional obfuscation. We should only ever need to do the little endian conversion before a `withUnsafeBytes` call and ideally we won't add any more direct calls to that and instead always go through `withUTF8` and `withMutableCapacity`.

mundaym replied:
We could also go back to the approach I took in 2866b4a and always store `_SmallString._storage` in little endian byte order. That avoids byte swapping before a read but means that `_StringObject` and `_SmallString` have different raw values for the same string. Though this is also true for 32-bit architectures, so maybe we should go down that path. It would have the nice side effect of making `withUnsafeBytes` always safe on `_SmallString` values.

milseman replied:
Yeah, I think that's the path we'll want to go down. For 64-bit LE, `_StringObject` and `_SmallString` have equivalent layout, but otherwise we need to do extra unpacking / byte-swapping as needed.

That will probably make further maintenance easier as well, as we're more likely to add more readers than creators.

milseman replied:
... and might let any SIMD, which we really want to do (on LE platforms at the very least), at least have a chance of being portable.

milseman reviewed (changes requested):
Why do we want to store SmallString as little endian in memory, and swap on read? Wouldn't we want SmallString to be the in-memory representation on both platforms?

mundaym commented on stdlib/public/core/SmallString.swift:
```diff
@@ -112,23 +116,13 @@ extension _SmallString {
-  // Give raw, nul-terminated code units. This is only for limited internal
-  // usage: it always clears the discriminator and count (in case it's full)
-  @inlinable
-  internal var zeroTerminatedRawCodeUnits: RawBitPattern {
```
I wasn't sure about this one. I think the new code is easier to understand without this property but I'm happy to reinstate it.

Why does this break the ABI? Is there any way to know when such breakages occur? This code isn't in a released version of Swift yet right?

milseman replied:
If you run the test suite, you'll get a failure inside the api-digester

mundaym replied:
I ran the validation test suite (specifically `utils/build-script --release-debuginfo -T`) on this PR on a Linux x86-64 box running Ubuntu 18.04 and it passed the tests. Should it have failed or will it only fail in the release branch?

Is the ABI still modifiable at this stage or is it already locked down?

milseman replied:
It is modifiable in case of emergency, and worse case here we just keep this method around.

mundaym commented on stdlib/public/core/SmallString.swift:
```diff
@@ -53,13 +53,17 @@ internal struct _SmallString {
+  internal init(_ object: _SmallString) {
+    self.init(raw: object.rawBits)
+  }
+
```
Oh, it does mean the code does an invariant check, whereas an assignment wouldn't I think.

milseman replied:
Right, if you want to do a final invariant check at the end. Alternatively, we could have the discriminator setter do the check. The discriminator property was added by @lorentey after I wrote much of this code, so there's some potential cleanups like the one you found.

lorentey replied:
Adding an invariant check to the discriminator setter could misfire in the ASCII check if the string contents haven't yet been fully updated at the time the discriminator is set.

milseman commented on stdlib/public/core/SmallString.swift:
```diff
@@ -245,41 +247,32 @@ extension _SmallString {
     let totalCount = base.count + other.count
     guard totalCount <= _SmallString.capacity else { return nil }
 
-    // TODO(SIMD): The below can be replaced with just be a couple vector ops
```
Right, but should Linux, macOS, and iOS suffer a several X performance regression in hot-paths to facilitate big-endian systems? At the very least, we would `#if` if necessary, assuming SIMD doesn't abstract it away sufficiently.

mundaym committed:
[string] Fix string implementation for big endian platforms

This commit supersedes 2866b4a which was overwritten by 4ab45df.

Store small string code units in little-endian byte order. This way
the code units are in the same order on all machines and can be
safely treated as an array of bytes.

mundaym commented:
I've reworked this PR to minimize the code changes and to remove the ABI changes. The code units in small strings are now always stored in little-endian byte order rather than being re-ordered when accessed as an array. Please take a look when you get a chance. Thanks!

milseman commented:
I’m a little confused. We had said:

> We could also go back to the approach I took in  [2866b4a](https://github.com/apple/swift/commit/2866b4a30d9e72188c760d109817fe7f9c4fd5b6)  and always store _SmallString._storage in little endian byte order. That avoids byte swapping before a read but means that _StringObject and _SmallString have different raw values for the same string. Though this is also true for 32-bit architectures, so maybe we should go down that path. It would have the nice side effect of making withUnsafeBytes always safe on _SmallString values.

My reading of that was that we would byte swap when converting between `_SmallString` and a `_StringObject` (analogous to 32-bit), rather than before every read. That would make it so that on any platform, the `_SmallString `’s in-memory representation is that of a 16-byte Array from first to last code unit, with a discriminator+count stored in the last byte. We’re much more likely to be extending/enhancing/maintaining readers of the contents of `_SmallString ` than the converter to/from `_StringObject`.

In that sense, `_StringObject` stores the contents in little-endian byte order so that the discriminator stays in the high bits available from our 50-bit limited address space. They get swapped when forming the `_SmallString` struct, wherein they are now stored in native memory order (i.e. pretend its an Array). This seems less invasive in code as well, and keeps consistency by having `_SmallString.rawBits` be the raw bits of a `_SmallString`, rather than byte-swapped.

A read from a `String` will have a byte-swap somewhere down the line, it just seems the cleanest for that to be during the `_StringObject` -> `_SmallString` conversion so that we can have uniform code that accesses the `_SmallString`, like on 32-bit. This also speeds up code that does multiple operations on the `_SmallString` by avoiding extra byte swaps. Is there a downside to this approach that I’m missing?

mundaym commented:
In the new code `_StringObject` is always stored in native byte order (so that the discriminator is in the right place as you mentioned). Then when we create a `_SmallString` we put the bytes into little-endian order. This happens in the `_SmallString` constructor rather than in `_StringObject` but the logic could live in either place. I think it's cleaner this way (only `_SmallString` needs to know it is in little-endian byte order).

The little-endian byte order of the raw bits in storage means that on big-endian systems the bytes in each 64-bit chunk of the stored tuple are reversed. This is great for operations such as `withUTF8` which will now work out of the box. It does, however, add some extra complexity to operations, such as `append`, that manipulate the stored code units using 64-bit operations. Either we create two code paths, one for big-endian systems and one for little-endian systems, or we convert the tuple into native byte order before manipulating it and convert it back into little-endian byte order. I went for the second approach since it means that the algorithm is the same on all platforms and we only need one code path. The cost of byte swaps is negligible.

Ideally I'd change the data type of the `RawBitPattern` (like I did in  2866b4a) to encapsulate the endianness of the platform completely, but that changes the ABI. This nice thing with the new code is that it shouldn't have any effect on little-endian platforms (since `UInt64.littleEndian` is a no-op), however it is a bit more fragile than the fully encapsulated solution.

milseman commented:
I feel like we’re talking past each other, and it might be due to conflicting terminology. Rather than say “it’s in little endian”, let’s use the term “memory order” to mean the order in memory that it would be laid out as if it were a 16-byte Array of UInt8 starting at the first code unit and finishing with the discriminator and count. To me, “memory order” seems like it would be equivalent to platform-native endianness, and I think we’re trying to say that `_SmallString`is in memory order.

We both agree that we do a byte swap on BE platforms when creating a `_SmallString` from a `_StringObject`. This puts the discriminator in the LSB of the second word, and the first byte is in the MSB of the first word. I.e., it’s now in memory order. I wouldn't think of this as “little endian”; if anything it seems like `_StringObject` is the one in little endian because its first code unit is in the LSB of the first word, and the discriminator is in the MSB of the second word.

Back to the code. I find it pretty confusing and brittle to change `rawBits` to no longer be the raw bits, but rather byte swap away from memory-order. That means that an initializer on `_SmallString` which wanted to take these raw bits would need to remember to byte swap them before storing them, i.e. they’re not really the raw bits. Furthermore, code such as the appending initializer needs to do another byte-swap to undo the prior byte-swap done by `rawBits`, and this is all very confusing. All of this byte swapping is mostly just there to facilitate subscript, because subscript uses a helper function that byte-indexes a UInt with 0 meaning the LSB. Wouldn’t it be clearer to instead change `_uncheckedGetByte`, `_uncheckedSetByte`, and `_bytesToUInt64` to be endianness correct, rather than byte swapping before and after every use? E.g. `_uncheckedGetByte` uses `self.littleEndian` rather than `self`.

mundaym commented:
Thanks for your patience. I've updated the PR with your suggestions. Is it a bit closer to what you had in mind now?

milseman commented on stdlib/public/core/SmallString.swift:
```diff
@@ -295,23 +312,27 @@ extension _SmallString {
-  //
-  // TODO: endianess awareness day
+  // Fetches the `i`th byte, from left to right.
   @inlinable @inline(__always)
```
What does "left to right" mean, memory-order/host-endianess?

mundaym replied:
Clarified the comment (memory-order).

milseman reviewed (approved):
This is looking great! Thanks for your patience as well.

Review notes are just details about comments, which I don't mind doing in a followup PR. I'll start testing and benchmarking now.

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please benchmark

mundaym committed:
[string] Improve some small string comments

Clarify the expected behavior of some small string helper functions.

mundaym commented on stdlib/public/core/SmallString.swift:
```diff
@@ -74,7 +89,8 @@ extension _SmallString {
 
   @inlinable @inline(__always)
```
Added a brief comment. Happy to drop this additional commit if you'd rather do a comment follow up yourself.

milseman commented:
@swift-ci please test

mundaym commented:
@milseman Is there a way to restart the CI tests? I'm not sure what went wrong with them but I don't *think* it was anything to do with this change. Thanks!

milseman commented:
@swift-ci please test os x platform

milseman commented:
@swift-ci please test linux platform

milseman commented:
@swift-ci please test os x platform
